🎉 Source Freshdesk: Migrated to latest CDK

lgomezm opened the pull request:
## What
* The Freshdesk source connector was written way before the current CDK version and is using outdated patterns.
* It should be re-written using the latest CDK patterns,

## How
* Migrated the source connector to latest CDK.
* Code is significantly simplified after these changes.
* Added a few tests to cover new code.

## Recommended reading order
1. `source.py`
2. `streams.py`
3. Files in `unit_tests`.

## 🚨 User Impact 🚨
Are there any breaking changes? What is the end result perceived by the user? If yes, please merge this PR with the 🚨🚨 emoji so changelog authors can further highlight this if needed.

## Pre-merge Checklist
Expand the relevant checklist and delete the others.

<details><summary><strong>New Connector</strong></summary>

### Community member or Airbyter

- [ ] **Community member?** Grant edit access to maintainers ([instructions](https://docs.github.com/en/github/collaborating-with-pull-requests/working-with-forks/allowing-changes-to-a-pull-request-branch-created-from-a-fork#enabling-repository-maintainer-permissions-on-existing-pull-requests))
- [ ] Secrets in the connector's spec are annotated with `airbyte_secret`
- [ ] Unit & integration tests added and passing. Community members, please provide proof of success locally e.g: screenshot or copy-paste unit, integration, and acceptance test output. To run acceptance tests for a Python connector, follow instructions in the README. For java connectors run `./gradlew :airbyte-integrations:connectors:<name>:integrationTest`.
- [ ] Code reviews completed
- [ ] Documentation updated
    - [ ] Connector's `README.md`
    - [ ] Connector's `bootstrap.md`. See [description and examples](https://docs.google.com/document/d/1ypdgmwmEHWv-TrO4_YOQ7pAJGVrMp5BOkEVh831N260/edit?usp=sharing)
    - [ ] `docs/SUMMARY.md`
    - [ ] `docs/integrations/<source or destination>/<name>.md` including changelog. See changelog [example](https://docs.airbyte.io/integrations/sources/stripe#changelog)
    - [ ] `docs/integrations/README.md`
    - [ ] `airbyte-integrations/builds.md`
- [ ] PR name follows [PR naming conventions](https://docs.airbyte.io/contributing-to-airbyte/updating-documentation#issues-and-pull-requests)

### Airbyter

If this is a community PR, the Airbyte engineer reviewing this PR is responsible for the below items.

- [ ] Create a non-forked branch based on this PR and test the below items on it
- [ ] Build is successful
- [ ] If new credentials are required for use in CI, add them to GSM. [Instructions](https://docs.airbyte.io/connector-development#using-credentials-in-ci).
- [ ] [`/test connector=connectors/<name>` command](https://docs.airbyte.io/connector-development#updating-an-existing-connector) is passing
- [ ] New Connector version released on Dockerhub by running the `/publish` command described [here](https://docs.airbyte.io/connector-development#updating-an-existing-connector)
- [ ] After the connector is published, connector added to connector index as described [here](https://docs.airbyte.io/connector-development#publishing-a-connector)
- [ ] Seed specs have been re-generated by building the platform and committing the changes to the seed spec files, as described [here](https://docs.airbyte.io/connector-development#publishing-a-connector)

</details>

<details><summary><strong>Updating a connector</strong></summary>

### Community member or Airbyter

- [ ] Grant edit access to maintainers ([instructions](https://docs.github.com/en/github/collaborating-with-pull-requests/working-with-forks/allowing-changes-to-a-pull-request-branch-created-from-a-fork#enabling-repository-maintainer-permissions-on-existing-pull-requests))
- [ ] Secrets in the connector's spec are annotated with `airbyte_secret`
- [ ] Unit & integration tests added and passing. Community members, please provide proof of success locally e.g: screenshot or copy-paste unit, integration, and acceptance test output. To run acceptance tests for a Python connector, follow instructions in the README. For java connectors run `./gradlew :airbyte-integrations:connectors:<name>:integrationTest`.
- [ ] Code reviews completed
- [ ] Documentation updated
    - [ ] Connector's `README.md`
    - [ ] Connector's `bootstrap.md`. See [description and examples](https://docs.google.com/document/d/1ypdgmwmEHWv-TrO4_YOQ7pAJGVrMp5BOkEVh831N260/edit?usp=sharing)
    - [ ] Changelog updated in `docs/integrations/<source or destination>/<name>.md` including changelog. See changelog [example](https://docs.airbyte.io/integrations/sources/stripe#changelog)
- [ ] PR name follows [PR naming conventions](https://docs.airbyte.io/contributing-to-airbyte/updating-documentation#issues-and-pull-requests)

### Airbyter

If this is a community PR, the Airbyte engineer reviewing this PR is responsible for the below items.

- [ ] Create a non-forked branch based on this PR and test the below items on it
- [ ] Build is successful
- [ ] If new credentials are required for use in CI, add them to GSM. [Instructions](https://docs.airbyte.io/connector-development#using-credentials-in-ci).
- [ ] [`/test connector=connectors/<name>` command](https://docs.airbyte.io/connector-development#updating-an-existing-connector) is passing
- [ ] New Connector version released on Dockerhub and connector version bumped by running the `/publish` command described [here](https://docs.airbyte.io/connector-development#updating-an-existing-connector)

</details>

<details><summary><strong>Connector Generator</strong></summary>

- [ ] Issue acceptance criteria met
- [ ] PR name follows [PR naming conventions](https://docs.airbyte.io/contributing-to-airbyte/updating-documentation#issues-and-pull-requests)
- [ ] If adding a new generator, add it to the [list of scaffold modules being tested](https://github.com/airbytehq/airbyte/blob/master/airbyte-integrations/connector-templates/generator/build.gradle#L41)
- [ ] The generator test modules (all connectors with `-scaffold` in their name) have been updated with the latest scaffold by running `./gradlew :airbyte-integrations:connector-templates:generator:testScaffoldTemplates` then checking in your changes
- [ ] Documentation which references the generator is updated as needed

</details>

## Tests

<details><summary><strong>Unit</strong></summary>

*Put your unit tests output here.*

</details>

<details><summary><strong>Integration</strong></summary>

*Put your integration tests output here.*

</details>

<details><summary><strong>Acceptance</strong></summary>

*Put your acceptance tests output here.*

</details>

lgomezm commented:
Here's the acceptance test results:
<img width="1407" alt="Screen Shot 2022-04-24 at 11 47 59 PM" src="https://user-images.githubusercontent.com/781929/165215561-5bb6a418-56ca-4b90-981c-4155d709f6d1.png">

lgomezm commented:
Here's the unit test results:
<img width="1405" alt="Screen Shot 2022-04-25 at 10 39 13 PM" src="https://user-images.githubusercontent.com/781929/165216148-009f3f06-4b01-48d4-9d68-f506e13dc467.png">

marcosmarxm commented:
/test connector=connectors/source-freshdesk

itaseskii commented:
I'll start code reviewing this.

alafanechere commented:
Thanks @itaseskii , feel free to ping me if you need an additional review.

itaseskii commented on airbyte-integrations/connectors/source-freshdesk/unit_tests/test_source.py:
```diff
@@ -0,0 +1,69 @@
+    requests_mock.register_uri("GET", "/api/v2/settings/helpdesk", responses)
+    ok, error_msg = SourceFreshdesk().check_connection(logger, config=config)
+
+    assert not ok
```
use single line statements/assertions `assert not ok and error_msg == ""`

lgomezm replied:
Updated in [4e2f437](https://github.com/airbytehq/airbyte/pull/12334/commits/4e2f4374641f906dfc9b9c69a77f02847186ef65).

itaseskii commented on airbyte-integrations/connectors/source-freshdesk/source_freshdesk/streams.py:
```diff
@@ -0,0 +1,273 @@
+        self, stream_state: Mapping[str, Any], stream_slice: Mapping[str, Any] = None, next_page_token: Mapping[str, Any] = None
+    ) -> Mapping[str, Any]:
+        return {
+            "Content-Type": "application/json",
```
Why is _**"Content-Type": "application/json"**_ needed when sending a request? Aren't all the requests of GET type and without a body?

lgomezm replied:
You're right. Turns out it works without having to send these headers. I simplified the code by removing this method in [4e2f437](https://github.com/airbytehq/airbyte/pull/12334/commits/4e2f4374641f906dfc9b9c69a77f02847186ef65).

itaseskii commented on airbyte-integrations/connectors/source-freshdesk/source_freshdesk/streams.py:
```diff
@@ -0,0 +1,273 @@
+        return "time_entries"
+
+
+class Tickets(IncrementalFreshdeskStream):
```
Given that **_Tickets_** will be retrieved in ascending order maybe we can set a **_state_checkpoint_interval = 100_** in order to persist state on every 100 records and avoid sending previously sent records if the replication crashes midway?

https://github.com/airbytehq/airbyte/blob/master/docs/connector-development/cdk-python/incremental-stream.md#interval-based-checkpointing

lgomezm replied:
Awesome. I added this in [4e2f437](https://github.com/airbytehq/airbyte/pull/12334/commits/4e2f4374641f906dfc9b9c69a77f02847186ef65). Thank you for the suggestion!

alafanechere replied:
I suggest you set `state_checkpoint_interval` on all the Incremental streams, or you can implement `stream_slices`. If you implement stream slicing the state will be checkpointed once a slice was properly consumed.

lgomezm replied:
Added in [abcae41](https://github.com/airbytehq/airbyte/pull/12334/commits/abcae4135576a3541a76979a09babf3a955c6b8f).

itaseskii commented on airbyte-integrations/connectors/source-freshdesk/source_freshdesk/streams.py:
```diff
@@ -0,0 +1,273 @@
+        pagination_complete = False
+
+        next_page_token = None
+        with AirbyteSentry.start_transaction("read_records", self.name), AirbyteSentry.start_transaction_span("read_records"):
```
Would you mind explaining why copying large amounts of the parent stream`read_records()`is needed instead of invoking `super().read_records()`

lgomezm replied:
You made me re-think how I migrated this particular stream and I ended up overriding the `next_page_token` method instead of the `read_records` one. In the end, I had to intervene the `next_page_token` if the next page to get was above 300. The method now has a comment that was ported from the original implementation. See [b0bcf0e](https://github.com/airbytehq/airbyte/pull/12334/commits/b0bcf0e00310eb439c80526604439211cd0adf97).

itaseskii commented:
Will do another round of code reviewing tomorrow.

lgomezm commented on airbyte-integrations/connectors/source-freshdesk/source_freshdesk/source.py:
```diff
@@ -2,11 +2,51 @@
-class SourceFreshdesk(BaseSource):
-    client_class = Client
+    def _create_authenticator(self, api_key: str) -> AuthBase:
+        return HTTPBasicAuth(username=api_key, password="unused_with_api_key")
```
You should provide a password. Otherwise you'll get an error like this:
```
$ python3 main.py check --config secrets/config.json                                     
{"type": "LOG", "log": {"level": "ERROR", "message": "Check failed"}}
{"type": "CONNECTION_STATUS", "connectionStatus": {"status": "FAILED", "message": "'TypeError(\"__init__() missing 1 required positional argument: \\'password\\'\")'"}}
```

alafanechere replied:
@lgomezm I would suggest creating a small child class from `HTTPBasicAuth` (such as `HTTPBasicAuthNoPassword`), in which you set an empty default password. This will make this part of the code easier to read and understand that Freshdesk is using a specific auth mechanism.

lgomezm replied:
Sure. I added that small class in [abcae41](https://github.com/airbytehq/airbyte/pull/12334/commits/abcae4135576a3541a76979a09babf3a955c6b8f).

lgomezm commented on airbyte-integrations/connectors/source-freshdesk/source_freshdesk/source.py:
```diff
@@ -2,11 +2,51 @@
+    def streams(self, config: Mapping[str, Any]) -> List[Stream]:
+        authenticator = self._create_authenticator(config["api_key"])
+        return [
+            Agents(authenticator=authenticator, config=config),
```
Good catch! I should have missed it when migrating individual streams. Added back in [4e2f437](https://github.com/airbytehq/airbyte/pull/12334/commits/4e2f4374641f906dfc9b9c69a77f02847186ef65).

lgomezm commented on airbyte-integrations/connectors/source-freshdesk/source_freshdesk/source.py:
```diff
@@ -2,11 +2,51 @@
+        alive = True
+        error_msg = None
+        try:
+            url = f"https://{config['domain'].rstrip('/')}/api/v2/settings/helpdesk"
```
Nice. I didn't know about that function. Changed in [4e2f437](https://github.com/airbytehq/airbyte/pull/12334/commits/4e2f4374641f906dfc9b9c69a77f02847186ef65).

alafanechere commented on airbyte-integrations/connectors/source-freshdesk/source_freshdesk/streams.py:
```diff
@@ -0,0 +1,273 @@
+            yield element
+
+
+class IncrementalFreshdeskStream(FreshdeskStream, ABC):
```
```suggestion
class IncrementalFreshdeskStream(FreshdeskStream, IncrementalMixin):
```
The latest CDK introduces the `IncrementalMixin` class. I suggest you use it and set the state value in `read_records` as suggested [here](https://docs.airbyte.com/connector-development/cdk-python/incremental-stream/#incrementalmixin). `get_updated_state` is now deprecated.

lgomezm replied:
Added in [abcae41](https://github.com/airbytehq/airbyte/pull/12334/commits/abcae4135576a3541a76979a09babf3a955c6b8f)

alafanechere commented on airbyte-integrations/connectors/source-freshdesk/source_freshdesk/source.py:
```diff
@@ -2,11 +2,51 @@
+        error_msg = None
+        try:
+            url = f"https://{config['domain'].rstrip('/')}/api/v2/settings/helpdesk"
+            r = requests.get(url=url, auth=self._create_authenticator(config["api_key"]))
```
```suggestion
            response = requests.get(url=url, auth=self._create_authenticator(config["api_key"]))
```
please avoid single-letter variable names.

lgomezm replied:
Updated in [abcae41](https://github.com/airbytehq/airbyte/pull/12334/commits/abcae4135576a3541a76979a09babf3a955c6b8f)

alafanechere commented on airbyte-integrations/connectors/source-freshdesk/source_freshdesk/source.py:
```diff
@@ -2,11 +2,51 @@
+        alive = True
+        error_msg = None
+        try:
+            url = f"https://{config['domain'].rstrip('/')}/api/v2/settings/helpdesk"
```
nit: a more robust way of checking connection would be to request several actual streams endpoints

alafanechere commented on airbyte-integrations/connectors/source-freshdesk/source_freshdesk/source.py:
```diff
@@ -2,11 +2,51 @@
+        try:
+            url = f"https://{config['domain'].rstrip('/')}/api/v2/settings/helpdesk"
+            r = requests.get(url=url, auth=self._create_authenticator(config["api_key"]))
+            if not r.ok:
```
I would suggest leveraging `r.raise_for_status()` and handling the error if it's raised.

lgomezm replied:
Sure. Updated in [24d1581](https://github.com/airbytehq/airbyte/pull/12334/commits/24d158184627496eabeb1cd9458dbc65f32a5f11).

alafanechere commented on airbyte-integrations/connectors/source-freshdesk/source_freshdesk/source.py:
```diff
@@ -2,11 +2,51 @@
+    
+    def streams(self, config: Mapping[str, Any]) -> List[Stream]:
+        authenticator = self._create_authenticator(config["api_key"])
+        return [
+            Agents(authenticator=authenticator, config=config),
```
nit:
```suggestion
        stream_kwargs = {"authenticator": authenticator, "config": config}
        return [
            Agents(**stream_kwargs),
```
This could ease future changes in the instantiation of streams.

lgomezm replied:
Nice suggestion. Applied in [abcae41](https://github.com/airbytehq/airbyte/pull/12334/commits/abcae4135576a3541a76979a09babf3a955c6b8f).

alafanechere commented on airbyte-integrations/connectors/source-freshdesk/source_freshdesk/streams.py:
```diff
@@ -0,0 +1,273 @@
+    ) -> Mapping[str, Any]:
+        return {
+            "Content-Type": "application/json",
+            "User-Agent": "Airbyte",
```
Why are we setting this user agent here? 
If this has no impact on Freshdesk I think it should not be set and it should be CDK's responsibility to set it.

lgomezm replied:
Yeah, I removed these headers as they were not needed.

alafanechere commented on airbyte-integrations/connectors/source-freshdesk/source_freshdesk/streams.py:
```diff
@@ -0,0 +1,273 @@
+from source_freshdesk.utils import CallCredit
+
+
+LINK_REGEX = re.compile(r'<(.*?)>;\s*rel="next"')
```
This can be a class attribute of `FreshdeskStream` if it's not used elsewhere.

lgomezm replied:
Sure, updated in [abcae41](https://github.com/airbytehq/airbyte/pull/12334/commits/abcae4135576a3541a76979a09babf3a955c6b8f).

alafanechere commented on airbyte-integrations/connectors/source-freshdesk/source_freshdesk/streams.py:
```diff
@@ -0,0 +1,273 @@
+    def parse_response(self, response: requests.Response, **kwargs) -> Iterable[Mapping]:
+        data = response.json()
+        if not data:
+            return
```
```suggestion
            return []
```
Please return an empty iterable to respect to return type.

lgomezm replied:
Done in [abcae41](https://github.com/airbytehq/airbyte/pull/12334/commits/abcae4135576a3541a76979a09babf3a955c6b8f).

lgomezm commented:
Hi @alafanechere. Thank you for reviewing. I've updated this PR as per your comments. Please take a look again when you get a chance.

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/airbytehq/airbyte?pullRequest=12334) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/airbytehq/airbyte?pullRequest=12334) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/airbytehq/airbyte?pullRequest=12334) it.</sub>

alafanechere commented on airbyte-integrations/connectors/source-freshdesk/source_freshdesk/source.py:
```diff
@@ -34,13 +34,14 @@ def check_connection(self, logger: logging.Logger, config: Mapping[str, Any]) ->
+            try:
+                body = error.response.json()
+                error_msg = f"{body.get('code')}: {body['message']}"
+            except ValueError:
+                error_msg = "Invalid credentials"
```
In which situation `ValueError` is raised here?
If whatever non HTTPError is raised is I don't think the cause would be `Invalid Credential`.
I think this try except block is redundant as non HttpError are watched by the higher level `except Exception as error:`

lgomezm replied:
You're right. I've removed the handling of this kind of error.

alafanechere commented on airbyte-integrations/connectors/source-freshdesk/source_freshdesk/source.py:
```diff
@@ -2,11 +2,65 @@
 
-from .client import Client
 
+class HTTPBasicAuthNoPassword(HTTPBasicAuth):
```
nit:
```suggestion
class FreshdeskAuth(HTTPBasicAuth):
```

lgomezm replied:
Done.

alafanechere commented on airbyte-integrations/connectors/source-freshdesk/source_freshdesk/streams.py:
```diff
@@ -0,0 +1,254 @@
+        # Since this logic rely not on updated tickets, it can break tickets dependant streams - conversations.
+        # So updated_since parameter will be always used in tickets streams. And start_date will be used too
+        # with default value 30 days look back.
+        self.start_date = pendulum.parse(start_date) if start_date else pendulum.now() - pendulum.duration(days=30)
```
```suggestion
        self.start_date = pendulum.parse(config["start_date"]) if config["start_date"] else pendulum.now() - pendulum.duration(days=30)
```

alafanechere commented on airbyte-integrations/connectors/source-freshdesk/source_freshdesk/streams.py:
```diff
@@ -0,0 +1,254 @@
+
+class FreshdeskStream(HttpStream, ABC):
+    """Basic stream API that allows to iterate over entities"""
+    call_credit = 1  # see https://developers.freshdesk.com/api/#embedding
```
This value never changes, I'd prefer to decrement from `1` rather than decrementing from this variable if it never changes.

lgomezm replied:
I double checked and it was actually being used with a different value in the `Tickets` stream. I've updated it accordingly.

alafanechere commented on airbyte-integrations/connectors/source-freshdesk/source_freshdesk/streams.py:
```diff
@@ -0,0 +1,254 @@
+        start_date = config["start_date"]
+        self.domain = config["domain"]
+        self._call_credit = CallCredit(balance=requests_per_minute) if requests_per_minute else None
+        # By default, only tickets that have been created within the past 30 days will be returned.
+        # Since this logic rely not on updated tickets, it can break tickets dependant streams - conversations.
+        # So updated_since parameter will be always used in tickets streams. And start_date will be used too
+        # with default value 30 days look back.
```
Defaulting to a 30 days look-back window might confuse users if they set a specific start date and get records older than this `start_date`

lgomezm replied:
I am not sure how to handle this use case. I basically ported this portion from the existing implementation. Since here I was just migrating to the new CDK, I was not paying much attention to these kind of details. Please let me know what the standard way of handling these cases is.

alafanechere commented on airbyte-integrations/connectors/source-freshdesk/source_freshdesk/streams.py:
```diff
@@ -0,0 +1,254 @@
+    
+    def backoff_time(self, response: requests.Response) -> Optional[float]:
+        if response.status_code == requests.codes.too_many_requests:
+            return float(response.headers.get("Retry-After"))
```
```suggestion
            return float(response.headers.get("Retry-After", 0))
```
You will have a `TypeError` if `Retry-After` is not in the headers: `float(None) -> TypeError`

lgomezm replied:
Good catch! Done.

alafanechere commented on airbyte-integrations/connectors/source-freshdesk/source_freshdesk/streams.py:
```diff
@@ -0,0 +1,254 @@
+            match = self.link_regex.search(link_header)
+            next_url = match.group(1)
+            params = parse.parse_qs(parse.urlparse(next_url).query)
+            return self.parse_link_params(link_query_params=params)
```
```suggestion
            return {"per_page": params['per_page'][0], "page": params['page'][0]}
```

lgomezm replied:
Done.

alafanechere commented on airbyte-integrations/connectors/source-freshdesk/source_freshdesk/streams.py:
```diff
@@ -0,0 +1,254 @@
+        except Exception as e:
+            raise KeyError(f"error parsing next_page token: {e}")
+    
+    def parse_link_params(self, link_query_params: Mapping[str, List[str]]) -> Mapping[str, Any]:
+        return {"per_page": link_query_params['per_page'][0], "page": link_query_params['page'][0]}
```
```suggestion
```

lgomezm replied:
Done.

alafanechere commented on airbyte-integrations/connectors/source-freshdesk/source_freshdesk/streams.py:
```diff
@@ -0,0 +1,254 @@
+        self, stream_state: Mapping[str, Any], stream_slice: Mapping[str, any] = None, next_page_token: Mapping[str, Any] = None
+    ) -> MutableMapping[str, Any]:
+        params = {"per_page": self.result_return_limit}
+        if next_page_token and "page" in next_page_token:
```
How could `page` not be in `next_page_token` if `next_page_token` is not None?

alafanechere reviewed (changes requested):
I went for another partial review but need to switch context. I'll review the rest asap.

alafanechere commented on airbyte-integrations/connectors/source-freshdesk/source_freshdesk/streams.py:
```diff
@@ -0,0 +1,254 @@
+
+    def __init__(self, authenticator: AuthBase, config: Mapping[str, Any], *args, **kwargs):
+        super().__init__(authenticator=authenticator, config=config, *args, **kwargs)
+        self._state = None
```
I don't think you should use `_.state` class attribute if you expose the `state` property. You should rather manage a `self._cursor_value` attribute

lgomezm replied:
Makes sense. Updated.

alafanechere commented on airbyte-integrations/connectors/source-freshdesk/source_freshdesk/streams.py:
```diff
@@ -0,0 +1,254 @@
+
+    @state.setter
+    def state(self, value: MutableMapping[str, Any]):
+        if "updated_at" in value:
+            self._state = value["updated_at"]
```
```suggestion
self._cursor_value = value.get(self.cursor_field, self.start_date)
```

lgomezm replied:
Done.

alafanechere commented on airbyte-integrations/connectors/source-freshdesk/source_freshdesk/streams.py:
```diff
@@ -0,0 +1,254 @@
+    
+    @property
+    def state(self) -> MutableMapping[str, Any]:
+        return {"updated_at": self._state} if self._state else {}
```
```suggestion
        return {self.cursor_field: self._cursor_value}
```

lgomezm replied:
Done.

alafanechere commented on airbyte-integrations/connectors/source-freshdesk/source_freshdesk/streams.py:
```diff
@@ -0,0 +1,254 @@
+    ) -> Iterable[Mapping[str, Any]]:
+        for record in super().read_records(sync_mode=sync_mode, cursor_field=cursor_field, stream_slice=stream_slice, stream_state=stream_state):
+            yield record
+            if not self.state or self.state["updated_at"] < record[self.cursor_field]:
+                self.state = record
```
```suggestion
             self._cursor_value = max(record[self.cursor_field], self._cursor_value)
```

lgomezm replied:
Done.

alafanechere commented on airbyte-integrations/connectors/source-freshdesk/source_freshdesk/streams.py:
```diff
@@ -0,0 +1,254 @@
+        self, stream_state: Mapping[str, Any], stream_slice: Mapping[str, any] = None, next_page_token: Mapping[str, Any] = None
+    ) -> MutableMapping[str, Any]:
+        params = super().request_params(stream_state=stream_state, stream_slice=stream_slice, next_page_token=next_page_token)
+        if "updated_at" in stream_state:
+            params[self.state_filter] = stream_state["updated_at"]
+        else:
+            params[self.state_filter] = self.start_date
```
```suggestion
        params[self.cursor_filter] = stream_state[self.cursor_field]
```

lgomezm replied:
Done.

alafanechere commented on airbyte-integrations/connectors/source-freshdesk/source_freshdesk/streams.py:
```diff
@@ -0,0 +1,254 @@
+
+class IncrementalFreshdeskStream(FreshdeskStream, IncrementalMixin):
+
+    state_filter = "updated_since"  # Name of filter that corresponds to the state
```
```suggestion
    cursor_filter = "updated_since" 
```

lgomezm replied:
Done.

alafanechere commented on airbyte-integrations/connectors/source-freshdesk/source_freshdesk/streams.py:
```diff
@@ -0,0 +1,254 @@
+
+
+class Contacts(IncrementalFreshdeskStream):
+    state_filter = "_updated_since"
```
```suggestion
    cursor_filter = "_updated_since"
```

lgomezm replied:
Done.

alafanechere commented on airbyte-integrations/connectors/source-freshdesk/source_freshdesk/streams.py:
```diff
@@ -0,0 +1,254 @@
+
+class Contacts(IncrementalFreshdeskStream):
+    state_filter = "_updated_since"
+    state_checkpoint_interval = 100
```
I think you can set this checkpoint interval at the `IncrementalFreshdeskStream` level

lgomezm replied:
Sounds good. Done.

alafanechere commented on airbyte-integrations/connectors/source-freshdesk/source_freshdesk/streams.py:
```diff
@@ -0,0 +1,254 @@
+
+class Tickets(IncrementalFreshdeskStream):
+    ticket_paginate_limit = 300
+    state_checkpoint_interval = 100
```
Please set this at the `IncrementalFreshdeskStream` level

lgomezm replied:
Sure. Done.

alafanechere commented on airbyte-integrations/connectors/source-freshdesk/source_freshdesk/streams.py:
```diff
@@ -0,0 +1,254 @@
+        params = super().request_params(stream_state=stream_state, stream_slice=stream_slice, next_page_token=next_page_token)
+        params.update({
+            "order_type": "asc",  # ASC order, to get the old records first
+            "order_by": "updated_at",
```
```suggestion
            "order_by": self.cursor_field,
```

lgomezm replied:
Done.

alafanechere commented on airbyte-integrations/connectors/source-freshdesk/source_freshdesk/streams.py:
```diff
@@ -0,0 +1,254 @@
+        self, stream_state: Mapping[str, Any], stream_slice: Mapping[str, any] = None, next_page_token: Mapping[str, Any] = None
+    ) -> MutableMapping[str, Any]:
+        params = super().request_params(stream_state=stream_state, stream_slice=stream_slice, next_page_token=next_page_token)
+        params.update({
```
Do you think this is can also be set in the `IncrementalFreshdeskStream` so that we have the same ordering behavior for all incremental streams?

lgomezm replied:
Not every API supports these parameters. Most of them only accept the pagination ones.

alafanechere commented on airbyte-integrations/connectors/source-freshdesk/source_freshdesk/streams.py:
```diff
@@ -0,0 +1,254 @@
+            "order_type": "asc",  # ASC order, to get the old records first
+            "order_by": "updated_at",
+        })
+        if next_page_token and "updated_since" in next_page_token:
```
```suggestion
        if next_page_token and self.cursor_filter in next_page_token:
```

lgomezm replied:
Done.

alafanechere commented on airbyte-integrations/connectors/source-freshdesk/source_freshdesk/streams.py:
```diff
@@ -0,0 +1,254 @@
+            "order_by": "updated_at",
+        })
+        if next_page_token and "updated_since" in next_page_token:
+            params["updated_since"] = next_page_token["updated_since"]
```
```suggestion
            params = {**params, **next_page_token}
```

lgomezm replied:
Done.

alafanechere commented on airbyte-integrations/connectors/source-freshdesk/source_freshdesk/streams.py:
```diff
@@ -0,0 +1,254 @@
+        return params
+    
+    def parse_link_params(self, link_query_params: Mapping[str, List[str]]) -> Mapping[str, Any]:
+        params = super().parse_link_params(link_query_params)
+        if "updated_since" in link_query_params:
+            params["updated_since"] = link_query_params["updated_since"][0]
+        return params
```
```suggestion
        params = super().parse_link_params(link_query_params)
        if self.cursor_filter in link_query_params:
            params[self.cursor_filter] = link_query_params[self.cursor_filter][0]
        return params
```

lgomezm replied:
Done.

alafanechere commented on airbyte-integrations/connectors/source-freshdesk/source_freshdesk/streams.py:
```diff
@@ -0,0 +1,254 @@
+            last_record_updated_at = response.json()[-1]["updated_at"]
+            last_record_updated_at = pendulum.parse(last_record_updated_at)
+            # updating request parameters with last_record state
+            next_page_token["updated_since"] = last_record_updated_at
+            del next_page_token["page"]
```
```suggestion
            next_page_token[self.cursor_filter] = last_record_updated_at
            next_page_token.pop(page)
```

lgomezm replied:
Awesome! Done

alafanechere commented on airbyte-integrations/connectors/source-freshdesk/source_freshdesk/streams.py:
```diff
@@ -0,0 +1,254 @@
+    """Notes and Replies"""
+    def __init__(self, authenticator: AuthBase, config: Mapping[str, Any], *args, **kwargs):
+        super().__init__(authenticator=authenticator, config=config, args=args, kwargs=kwargs)
+        self.tickets_stream = Tickets(authenticator=authenticator, config=config)
```
If you don't want to make `Conversations` stream incremental in this PR (it can be incremental too as `Tickets` is incremental), please at least enable `use_cache = True` on `Tickets`

lgomezm replied:
Yeah, done.

alafanechere commented on airbyte-integrations/connectors/source-freshdesk/source_freshdesk/streams.py:
```diff
@@ -0,0 +1,254 @@
+
+
+class SatisfactionRatings(IncrementalFreshdeskStream):
+    state_filter = "created_since"
```
```suggestion
    cursor_filter = "created_since"
```

lgomezm replied:
Done.

alafanechere reviewed (changes requested):
I'm done reviewing `source.py` and `streams.py` :) I made a lot of suggestions to simplify and increase readability + match CDK best practices.

lgomezm commented on airbyte-integrations/connectors/source-freshdesk/source_freshdesk/source.py:
```diff
@@ -2,11 +2,65 @@
+        return alive, error_msg
+    
+    def streams(self, config: Mapping[str, Any]) -> List[Stream]:
+        authenticator = self._create_authenticator(config["api_key"])
```
Renamed authenticator class with `FreshdeskAuth` and used here.

lgomezm commented on airbyte-integrations/connectors/source-freshdesk/source_freshdesk/streams.py:
```diff
@@ -0,0 +1,254 @@
+            params = parse.parse_qs(parse.urlparse(next_url).query)
+            return self.parse_link_params(link_query_params=params)
+        except Exception as e:
+            raise KeyError(f"error parsing next_page token: {e}")
```
Yeah, I updated this method and removed the exception handling.

lgomezm commented on airbyte-integrations/connectors/source-freshdesk/source_freshdesk/streams.py:
```diff
@@ -0,0 +1,254 @@
+        requests_per_minute = config["requests_per_minute"]
+        start_date = config["start_date"]
+        self.domain = config["domain"]
+        self._call_credit = CallCredit(balance=requests_per_minute) if requests_per_minute else None
```
Yeah, this was also ported from the existing implementation. I am not sure if we should get rid of this `CallCredit` mechanism.

lgomezm commented on airbyte-integrations/connectors/source-freshdesk/source_freshdesk/streams.py:
```diff
@@ -0,0 +1,254 @@
+        if not data:
+            return []
+        for element in data:
+            yield element
```
Done.

lgomezm commented:
@alafanechere I've addressed your comments. Please take a look again when you get a chance and let me know if there's anything else you consider I should update. Thanks!

alafanechere commented:
/test connector=connectors/source-freshdesk

alafanechere commented:
Thank you for the changes @lgomezm , I'm running the test now and will go for a final review asap.

alafanechere commented:
/test connector=connectors/source-freshdesk

lgomezm commented:
Hi @alafanechere. I've pushed a change that fixes acceptance tests. Please give it a try again when you get a chance. Here's the results of them running locally:
<img width="1406" alt="Screen Shot 2022-05-19 at 10 39 25 PM" src="https://user-images.githubusercontent.com/781929/169446037-a361b442-1edf-4f39-b7d5-555bd7b3fd0d.png">

alafanechere commented:
/test connector=connectors/source-freshdesk

alafanechere commented:
We are facing an error while running the acceptance test with our Freshdesk account: 
```
The Skill Based Round Robin feature(s) is/are not supported in your plan. Please upgrade your account to use it.", "internal_message": "403 Client Error: Forbidden for url: https://newaccount1603334233301.freshdesk.com/api/skills?per_page=100
```

I'm trying to remove the `Skills` stream from the tests.

alafanechere commented:
/test connector=connectors/source-freshdesk

alafanechere commented:
@lgomezm I'll be off until next Monday. Could you please try to figure out what could be problematic with the `Skills` stream and try to ignore it from the integration tests if needed (it looks like my latest attempt in https://github.com/airbytehq/airbyte/pull/12334/commits/77ea41b18a125fc4a304d3b34d1bb384848b76db did not work)?

lgomezm commented:
Hi @alafanechere. It looks like the test account that's used to run integration tests is not allowed to get skills. I have overwritten the configured catalog file so that the `skills` stream is not included in the acceptance tests. They should pass now.

alafanechere commented:
/test connector=connectors/source-freshdesk

alafanechere commented:
Hey @lgomezm while I was off our team upgraded our freshdesk account to overcome the problem on the `skills` account (https://github.com/airbytehq/airbyte/pull/13158)

alafanechere commented:
/publish connector=connectors/source-freshdesk

alafanechere reviewed (approved):
Thank you for the contribution @lgomezm !
